### PR TITLE
Serve json schemas

### LINKED ISSUE #86: Schemas should be mirrored by hive at some well-known url

### PR DESCRIPTION
**Pull request by zepatrik**
## Related issue

closes #86 

## Proposed changes

All identity traits schemas have to be configured using a human readable ID and the corresponding URL. This PR enables multiple schemas to be used next to the default schema.
It also adds the `kratos.public/schemas/:id` endpoint that mirrors all schemas.

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md)
- [x] I have read the [security policy](../security/policy)
- [x] I confirm that this pull request does not address a security vulnerability. If this pull request addresses a security
vulnerability, I confirm that I got green light (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation within the code base (if appropriate)
- [ ] I have documented my changes in the [developer guide](https://github.com/ory/docs) (if appropriate)

### CODE REVIEW

File: identity/handler_test.go [hunk 1]
```diff
@@ -110,14 +124,14 @@ func TestHandler(t *testing.T) {
 		i.ID = x.ParseUUID(res.Get("id").String())
 		assert.EqualValues(t, "baz", res.Get("traits.bar").String(), "%s", res.Raw)
 		assert.Empty(t, res.Get("credentials").String(), "%s", res.Raw)
-		assert.EqualValues(t, viper.GetString(configuration.ViperKeyDefaultIdentityTraitsSchemaURL), res.Get("traits_schema_url").String(), "%s", res.Raw)
+		assert.EqualValues(t, defaultSchemaExternalURL, res.Get("traits_schema_url").String(), "%s", res.Raw)
```
Comment (aeneasr) [inline identity/handler_test.go]: Check if the ID is set as well

File: identity/handler_test.go [hunk 2]
```diff
@@ -110,14 +124,14 @@ func TestHandler(t *testing.T) {
...
 	})
 
 	t.Run("case=should be able to get the identity", func(t *testing.T) {
 		res := get(t, "/identities/"+i.ID.String(), http.StatusOK)
 		assert.EqualValues(t, i.ID.String(), res.Get("id").String(), "%s", res.Raw)
 		assert.EqualValues(t, "baz", res.Get("traits.bar").String(), "%s", res.Raw)
-		assert.EqualValues(t, viper.GetString(configuration.ViperKeyDefaultIdentityTraitsSchemaURL), res.Get("traits_schema_url").String(), "%s", res.Raw)
+		assert.EqualValues(t, defaultSchemaExternalURL, res.Get("traits_schema_url").String(), "%s", res.Raw)
```
Comment (aeneasr) [inline identity/handler_test.go]: Check if the ID is set as well

File: identity/identity.go [hunk 3]
```diff
@@ -35,10 +37,11 @@ type (
...
 
-		// TraitsSchemaURL is the JSON Schema to be used for validating the identity's traits.
-		//
-		// format: uri
-		TraitsSchemaURL string `json:"traits_schema_url,omitempty" faker:"-" db:"traits_schema_url"`
+		// TraitsSchemaID is the JSON Schema to be used for validating the identity's traits.
+		TraitsSchemaID string `json:"traits_schema_id" faker:"-" db:"traits_schema_id"`
+
+		// TraitsSchemaURL is the kratos public endpoint where the identity's traits schema can be fetched.
+		TraitsSchemaURL string `json:"traits_schema_url" faker:"-" db:"-"`
```
Comment (aeneasr) [inline identity/identity.go]: Add `format: uri`

File: identity/identity.go [hunk 4]
```diff
@@ -35,10 +37,11 @@ type (
 		// Credentials represents all credentials that can be used for authenticating this identity.
 		Credentials map[CredentialsType]Credentials `json:"-" faker:"-" db:"-"`
 
-		// TraitsSchemaURL is the JSON Schema to be used for validating the identity's traits.
-		//
-		// format: uri
-		TraitsSchemaURL string `json:"traits_schema_url,omitempty" faker:"-" db:"traits_schema_url"`
+		// TraitsSchemaID is the JSON Schema to be used for validating the identity's traits.
```
Comment (aeneasr) [inline identity/identity.go]:
```suggestion
		// TraitsSchemaID is the ID of the JSON Schema to be used for validating the identity's traits.
```

File: identity/identity.go [hunk 5]
```diff
@@ -35,10 +37,11 @@ type (
...
+		TraitsSchemaID string `json:"traits_schema_id" faker:"-" db:"traits_schema_id"`
+
+		// TraitsSchemaURL is the kratos public endpoint where the identity's traits schema can be fetched.
```
Comment (aeneasr) [inline identity/identity.go]:
```suggestion
		// TraitsSchemaURL is the URL of then endpoint where the identity's traits schema can be fetched from.
```

File: identity/pool.go [hunk 6]
```diff
@@ -59,20 +59,26 @@ type (
 func TestPool(p Pool) func(t *testing.T) {
 	return func(t *testing.T) {
 		viper.Set(configuration.ViperKeyDefaultIdentityTraitsSchemaURL, "file://./stub/identity.schema.json")
+		viper.Set(configuration.ViperKeyURLsSelfPublic, "http://example.com")
+		const altSchema = "altSchema"
+		viper.Set(configuration.ViperKeyIdentityTraitsSchemas, []map[string]string{{
```
Comment (aeneasr) [inline identity/pool.go]: Replace this with the struct as discussed

File: identity/pool.go [hunk 7]
```diff
@@ -123,13 +129,13 @@ func TestPool(p Pool) func(t *testing.T) {
...
 		t.Run("case=create and keep set values", func(t *testing.T) {
-			expected := passwordIdentity("file://./stub/identity-2.schema.json", "id-2")
+			expected := passwordIdentity(altSchema, "id-2")
 			require.NoError(t, p.CreateIdentity(context.Background(), expected))
 			createdIDs = append(createdIDs, expected.ID)
 
 			actual, err := p.GetIdentity(context.Background(), expected.ID)
 			require.NoError(t, err)
-			assert.Equal(t, "file://./stub/identity-2.schema.json", actual.TraitsSchemaURL)
+			assert.Equal(t, altSchema, actual.TraitsSchemaID)
```
Comment (aeneasr) [inline identity/pool.go]: Check the URL

File: identity/pool.go [hunk 8]
```diff
@@ -107,7 +113,7 @@ func TestPool(p Pool) func(t *testing.T) {
 			require.NoError(t, err)
 
 			assert.Equal(t, expected.ID, actual.ID)
-			assert.Equal(t, "file://./stub/identity.schema.json", actual.TraitsSchemaURL)
+			assert.Equal(t, configuration.DefaultIdentityTraitsSchemaID, actual.TraitsSchemaID)
```
Comment (aeneasr) [inline identity/pool.go]: Check the URL

File: identity/pool.go [hunk 9]
```diff
@@ -207,7 +213,7 @@ func TestPool(p Pool) func(t *testing.T) {
 			require.NoError(t, p.CreateIdentity(context.Background(), initial))
 			createdIDs = append(createdIDs, initial.ID)
 
-			assert.Equal(t, "file://./stub/identity.schema.json", initial.TraitsSchemaURL)
+			assert.Equal(t, configuration.DefaultIdentityTraitsSchemaID, initial.TraitsSchemaID)
```
Comment (aeneasr) [inline identity/pool.go]: Check the URL

File: identity/pool.go [hunk 10]
```diff
@@ -216,15 +222,15 @@ func TestPool(p Pool) func(t *testing.T) {
...
+			toUpdate.TraitsSchemaID = altSchema
 
 			err := p.UpdateIdentity(context.Background(), toUpdate)
 			require.Error(t, err)
 			assert.Contains(t, fmt.Sprintf("%+v", err), "A field was modified that updates one or more credentials-related settings.")
 
 			actual, err := p.GetIdentityConfidential(context.Background(), toUpdate.ID)
 			require.NoError(t, err)
-			assert.Equal(t, "file://./stub/identity.schema.json", actual.TraitsSchemaURL)
+			assert.Equal(t, configuration.DefaultIdentityTraitsSchemaID, actual.TraitsSchemaID)
```
Comment (aeneasr) [inline identity/pool.go]: Check the URL

File: identity/pool.go [hunk 11]
```diff
@@ -234,7 +240,7 @@ func TestPool(p Pool) func(t *testing.T) {
 			require.NoError(t, p.CreateIdentity(context.Background(), initial))
 			createdIDs = append(createdIDs, initial.ID)
 
-			assert.Equal(t, "file://./stub/identity.schema.json", initial.TraitsSchemaURL)
+			assert.Equal(t, configuration.DefaultIdentityTraitsSchemaID, initial.TraitsSchemaID)
```
Comment (aeneasr) [inline identity/pool.go]: Check the URL

File: identity/pool.go [hunk 12]
```diff
@@ -243,12 +249,12 @@ func TestPool(p Pool) func(t *testing.T) {
...
 			})
 			expected.Traits = Traits(`{"update":"me"}`)
-			expected.TraitsSchemaURL = "file://./stub/identity-2.schema.json"
+			expected.TraitsSchemaID = altSchema
 			require.NoError(t, p.UpdateIdentityConfidential(context.Background(), expected))
 
 			actual, err := p.GetIdentityConfidential(context.Background(), expected.ID)
 			require.NoError(t, err)
-			assert.Equal(t, "file://./stub/identity-2.schema.json", actual.TraitsSchemaURL)
+			assert.Equal(t, altSchema, actual.TraitsSchemaID)
```
Comment (aeneasr) [inline identity/pool.go]: Check the URL

File: identity/validator_test.go [hunk 13]
```diff
@@ -50,9 +50,18 @@ func TestSchemaValidator(t *testing.T) {
 	ts := httptest.NewServer(router)
 	defer ts.Close()
 
-	conf, _ := internal.NewRegistryDefault(t)
+	_, reg := internal.NewRegistryDefault(t)
 	viper.Set(configuration.ViperKeyDefaultIdentityTraitsSchemaURL, ts.URL+"/schema/firstName")
-	v := NewValidator(conf)
+	viper.Set(configuration.ViperKeyIdentityTraitsSchemas, []map[string]string{
```
Comment (aeneasr) [inline identity/validator_test.go]: Use the struct as discussed

File: identity/validator_test.go [hunk 14]
```diff
@@ -50,9 +50,18 @@ func TestSchemaValidator(t *testing.T) {
...
+		{
+			"id":  "whatever",
+			"url": ts.URL + "/schema/whatever",
+		}, {
+			"id":  "unreachable-url",
+			"url": ts.URL,
```
Comment (aeneasr) [inline identity/validator_test.go]: Make this explicit with `+"/i-do-not-exist"`

File: identity/validator_test.go [hunk 15]
```diff
@@ -77,31 +86,32 @@ func TestSchemaValidator(t *testing.T) {
...
-		{
-			i: &Identity{
-				TraitsSchemaURL: "not-a-url",
-				Traits:          Traits(`{ "firstName": "first-name", "lastName": "last-name", "age": 1 }`),
+				TraitsSchemaID: "unreachable-url",
+				Traits:         Traits(`{ "firstName": "first-name", "lastName": "last-name", "age": 1 }`),
 			},
 			err: "An internal server error occurred, please contact the system administrator",
 		},
+		// This test case is probably not useful because all schemas have to be configured beforehand
```
Comment (aeneasr) [inline identity/validator_test.go]: clean this up

File: persistence/sql/persister_identity.go [hunk 16]
```diff
@@ -258,3 +281,15 @@ func (p *Persister) validateIdentity(i *identity.Identity) error {
...
 	return nil
 }
+
+func (p *Persister) injectTraitsSchemaURL(i *identity.Identity) error {
+	s, err := p.r.IdentityTraitsSchemas().GetByID(i.TraitsSchemaID)
+	if err != nil {
+		return errors.WithStack(
+			herodot.ErrInternalServerError.
+				WithReasonf(
+					"Could not find the schema %s for this identities traits.", i.TraitsSchemaID))
```
Comment (aeneasr) [inline persistence/sql/persister_identity.go]:
```suggestion
					`The JSON Schema "%s" for this identity's traits could not be found.", i.TraitsSchemaID))
```

File: schema/handler.go [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+	if err != nil {
+		h.r.Writer().WriteError(w, r, err)
+		return
+	}
+
+	if s.URL.Scheme == "file://" {
+		f, err := os.Open(s.URL.Path)
+		if err != nil {
+			if os.IsNotExist(err) {
+				h.r.Writer().WriteError(w, r, herodot.ErrNotFound)
```
Comment (aeneasr) [inline schema/handler.go]:
```suggestion
				h.r.Writer().WriteError(w, r, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("The file for this JSON Schema ID could not be found or opened. This is a configuration issue.").WithDebugf("%+v", err)))
```

File: schema/handler.go [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+	s, err := h.r.IdentityTraitsSchemas().GetByID(ps.ByName("id"))
+	if err != nil {
+		h.r.Writer().WriteError(w, r, err)
+		return
+	}
+
+	if s.URL.Scheme == "file://" {
+		f, err := os.Open(s.URL.Path)
+		if err != nil {
+			if os.IsNotExist(err) {
```
Comment (aeneasr) [inline schema/handler.go]: You can leave this out

File: schema/handler.go [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+		}
+	} else {
+		resp, err := http.Get(s.URL.String())
+		if err != nil {
+			h.r.Writer().WriteError(w, r, err)
+			return
+		}
+		defer resp.Body.Close()
+
+		if _, err = io.Copy(w, resp.Body); err != nil {
```
Comment (aeneasr) [inline schema/handler.go]: Move this `io.Copy` outside of the if - use `var r io.ReaderCloser`

File: schema/handler.go [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+	}
+}
+
+const SchemasPath string = "schemas"
+
+func (h *Handler) RegisterPublicRoutes(public *x.RouterPublic) {
+	public.GET(fmt.Sprintf("/%s/:id", SchemasPath), h.get)
+}
+
+func (h *Handler) get(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
```
Comment (aeneasr) [inline schema/handler.go]: Write a test for this

File: schema/schema.go [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+	"github.com/ory/kratos/driver/configuration"
+
+	"github.com/pkg/errors"
+
+	"github.com/ory/x/urlx"
+)
+
+type Schemas []Schema
+
+func (s Schemas) GetByID(id string) (*Schema, error) {
```
Comment (aeneasr) [inline schema/schema.go]: Write a test for this

File: selfservice/flow/login/hook_test.go [hunk 22]
```diff
@@ -33,10 +33,12 @@ type mockPostHook struct {
 	modifyIdentity bool
 }
 
+var updatedSchema = map[string]string{"id": "updatedSchema", "url": "file://./stub/updated.schema.json"}
```
Comment (aeneasr) [inline selfservice/flow/login/hook_test.go]: struct

File: selfservice/strategy/oidc/strategy.go [hunk 23]
```diff
@@ -377,7 +377,7 @@ func (s *Strategy) processRegistration(w http.ResponseWriter, r *http.Request, a
 		return
 	}
 
-	i := identity.NewIdentity(s.c.DefaultIdentityTraitsSchemaURL().String())
+	i := identity.NewIdentity("")
```
Comment (aeneasr) [inline selfservice/strategy/oidc/strategy.go]: Wherever default is being used - use the default key to improve readability

File: identity/identity.go (hunk 3 above)
Comment (zepatrik) [reply identity/identity.go]: Uniform Resource Locator (URL)
Uniform Resource Identifier (URI)
Uniform Resource Name (URN)

as we need the location it should be `url` I think

**Pull request merged**